Consolidate sym/non-sym overloads for _make_wrapper_subclass

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114236

I'm not sure why we needed two overloads previously, let's find out! Removing the int overload is load bearing because it now forces specialization on SymInt arguments instead of falling through to the SymInt overload, see new test.

I decided NOT to allow storage offset simultaneously with None strides.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Consolidate sym/non-sym overloads for _make_wrapper_subclass

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

ezyang committed:
Update on "Consolidate sym/non-sym overloads for _make_wrapper_subclass"


Recommend review with whitespace turned off. I'm not sure why we needed two overloads previously, let's find out! Removing the int overload is load bearing because it now forces specialization on SymInt arguments instead of falling through to the SymInt overload.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -744,13 +719,19 @@ static PyObject* THPVariable_make_wrapper_subclass(
+      tensor_impl->set_sizes_and_strides(
+          sym_sizes, sym_strides.value(), sym_storage_offset);
+    } else {
+      TORCH_INTERNAL_ASSERT(!sym_storage_offset.has_value());
```
TORCH_CHECK since this is user facing API-ish

albanD commented on test/test_python_dispatch.py:
```diff
@@ -750,7 +750,7 @@ def test_storage(self) -> None:
         x = LoggingTensor(torch.ones(1))
         storage = x.untyped_storage()
-        self.assertRaises(RuntimeError, lambda: storage.data_ptr())
+        storage.data_ptr()
```
Should we expect this to always be 0?

ezyang replied:
yeah, I think so

ezyang committed:
Update on "Consolidate sym/non-sym overloads for _make_wrapper_subclass"


I'm not sure why we needed two overloads previously, let's find out! Removing the int overload is load bearing because it now forces specialization on SymInt arguments instead of falling through to the SymInt overload, see new test.

I decided NOT to allow storage offset simultaneously with None strides.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Consolidate sym/non-sym overloads for _make_wrapper_subclass"


I'm not sure why we needed two overloads previously, let's find out! Removing the int overload is load bearing because it now forces specialization on SymInt arguments instead of falling through to the SymInt overload, see new test.

I decided NOT to allow storage offset simultaneously with None strides.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -702,39 +697,39 @@ static PyObject* THPVariable_make_wrapper_subclass(
+        size_bytes,
         /*allocator=*/c10::GetAllocator(c10::kMeta),
         /*resizable=*/true};
+    // TODO: constructor should probably accept data pointer
```
nit: could you remove the for_blob comment above that is really outdated at this point?

ezyang replied:
ok will do when ci is all green

albanD reviewed (commented):
Nice fix! It does fix a bunch of xfail!

ezyang committed:
Update on "Consolidate sym/non-sym overloads for _make_wrapper_subclass"


I'm not sure why we needed two overloads previously, let's find out! Removing the int overload is load bearing because it now forces specialization on SymInt arguments instead of falling through to the SymInt overload, see new test.

I decided NOT to allow storage offset simultaneously with None strides.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Consolidate sym/non-sym overloads for _make_wrapper_subclass"


I'm not sure why we needed two overloads previously, let's find out! Removing the int overload is load bearing because it now forces specialization on SymInt arguments instead of falling through to the SymInt overload, see new test.

I decided NOT to allow storage offset simultaneously with None strides.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "comment fix only"
